enhancements/helm3/namespace-scoped-helm-repository-crd.md

```diff
@@ -0,0 +1,196 @@
+
+### Goals
+
+- New namespace-scoped CRD created for Helm repositories: ProjectHelmChartRepository
+- Able to create the new namespace-scoped custom resource, ProjectHelmChartRepository
+- Document in helm section describing the difference between ProjectHelmChartRepository and HelmChartRepository
+
+### Non-Goals
+
+- Change the existing cluster-scoped CRD, HelmChartRepository
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an OpenShift developer, I would like to be able to add Helm repositories within my project without asking for an administrator's permissions
+
+### API Extensions
+
+#### Introducing a new namespace-scoped custom resource, ProjectHelmChartRepository
+
+Example definition:
+```yaml
+apiVersion: helm.openshift.io/v1beta1
+kind: ProjectHelmChartRepository
+metadata:
+  name: my-enterprise-chart-repo
+spec:
+  url: https://my.chart-repo.org/stable
+
+  # optional and only needed for UI purposes
+  displayName: myChartRepo
+
+  # optional and only needed for UI purposes
+  description: my private chart repo
+```
+
+To add a new namespace-scoped Helm repository to a desired namespace:
+```shell
+$ cat <<EOF | oc apply --namespace my-namespace -f -
+apiVersion: helm.openshift.io/v1beta1
+kind: ProjectHelmChartRepository
+metadata:
+  name: stable
+spec:
+  url: https://kubernetes-charts-incubator.storage.googleapis.com
+  displayName: Public Helm stable charts
+  description: Public Helm stable charts hosted on HelmHub
+EOF
+
+$ kubectl get projecthelmchartrepositories --namespace my-namespace
+NAME          AGE
+stable        1m
+```
+
+The addition of namespace-scoped Helm repository does not impact behavior of the existing cluster-scoped Helm repository:
+
+```shell
+$ cat <<EOF | oc apply -f -
+apiVersion: helm.openshift.io/v1beta1
+kind: HelmChartRepository
+metadata:
+  name: stable
+spec:
+  url: https://kubernetes-charts.storage.googleapis.com
+  displayName: Public Helm stable charts
+  description: Public Helm stable charts hosted on HelmHub
+EOF
+
+$ cat <<EOF | oc apply --namespace my-namespace -f -
+apiVersion: helm.openshift.io/v1beta1
+kind: ProjectHelmChartRepository
+metadata:
+  name: incubator
+spec:
+  url: https://kubernetes-charts-incubator.storage.googleapis.com
+  displayName: Public Helm charts in incubator state
+EOF
+
+$ kubectl get helmchartrepositories
+NAME          AGE
+cluster       3h30m
+stable        1m
+
+$ kubectl get projecthelmchartrepositories --namespace my-namespace
+NAME          AGE
+incubator     1m
+```
+
+#### How would the UI render the namespace-scoped Helm repository
+
+As defined in [previous helm enhancement](https://github.com/openshift/enhancements/blob/master/enhancements/helm3/console.md#how-would-the-ui-discover-the-charts), the UI will invoke the `/api/helm/charts/index.yaml` endpoint and the endpoint will proxy requests to the configured chart repositories.
```

Comment:
hm, so will /api/help/charts/index.yaml do any filtering/rbac?  or everyone will see the charts provided by any ProjectHelmChartRepository resource defined in any namespace in the cluster?

Reply:
The dev console backend will authenticate request with token: https://github.com/openshift/console/blob/64f95b6f5338950765a0b1b7dc6eb866757207eb/pkg/auth/auth_openshift.go#L176

However, you are correct on this assumption, the `/api/help/charts/index.yaml` API will fetch whatever the user has requested as long as the token is valid, including the Helm repositories in other namespaces.

However in implementation, the openshift-console will be the one sending this request. So the backend will not know the original user of the request (the human that is accessing the dev console). We might want to add another query parameter like `?user=user1` to return _only_ the helm repositories that are available to this `user1`, not any helm repository `user1` requested.

Reply:
Need some insights from dev console team members as well and any suggestion is welcomed.

cc @dperaza4dustbit @christianvogt @rohitkrai03

Reply:
I assumed we would provide `namespace=<namespace` query parameter to the api call and therefore only receive helm charts which are both cluster and namespace scoped. If no namespace supplied, then only cluster scoped resources are returned.

From the console backend you should have access to the user which invoked the console backend api.

cc @serenamarie125

Reply:
> I assumed we would provide namespace=<namespace query parameter to the api call and therefore only receive helm charts which are both cluster and namespace scoped. If no namespace supplied, then only cluster scoped resources are returned.

Agreed, this aligns with my understanding.

I thought twice about the RBAC, please correct me if I'm wrong:

For the user RBAC, no extra RBAC policy is added in this enhancement. I would assume the RBAC is done purely based on the bearer token in the HTTP request and if the token/user/role of user does not have access to the CRD in the requested namespace, then the resource will not be responsed back.

Reply:
Yes, this makes sense.

Reply:
@zonggen should it mentioned that `/api/helm/charts/index.yaml` needs to pass also namespace as the argument, for example:

* `/api/helm/charts/index.yaml?namespace=foo` or
* `/api/helm/charts/foo/index.yaml`

Reply:
@pedjak Thanks for the comments. I've made further clarifications in this section on the URL parameter `namespace`.